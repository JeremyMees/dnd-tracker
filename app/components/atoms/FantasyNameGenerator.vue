<script setup lang="ts">
import { useToast } from '~/components/ui/toast/use-toast'

const props = defineProps<{ amount: number }>()

const { copy } = useClipboard()
const { toast } = useToast()
const { t } = useI18n()

const names = ref<string[]>([])

onMounted(() => generate())

function generate(): void {
  names.value = []

  for (let i = 0; i < props.amount; i++) {
    names.value.push(randomName())
  }
}

function handleCopy(name: string): void {
  copy(name)

  toast({
    title: t('components.nameGenerator.toast.title', { name }),
    variant: 'info',
  })
}
</script>

<template>
  <MasonryGrid
    v-if="names.length"
    v-slot="{ column }"
    :data="names"
    :max-columns="2"
    wrapper-style="grid list-disc list-inside gap-x-6"
    column-style="flex flex-col gap-1"
    element="ol"
  >
    <li
      v-for="name in column"
      :key="name"
      class="cursor-copy"
      @click="handleCopy(name)"
    >
      {{ name }}
    </li>
  </MasonryGrid>
  <SkeletonList
    v-else
    :amount="amount"
  />

  <div class="flex w-full justify-end gap-2 items-center mt-4">
    <p class="text-muted-foreground text-2xs">
      {{ $t('pages.fantasyNameGenerator.tip') }}
    </p>
    <button
      data-test-generate
      :disabled="!names.length"
      class="btn-primary"
      @click="generate"
    >
      {{ $t('actions.generate') }}
    </button>
  </div>
</template>
