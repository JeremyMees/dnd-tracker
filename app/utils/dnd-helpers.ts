import { initiativeKeys } from '~~/constants/dnd-rules'
import { firstName, lastName, middleName } from '~~/constants/names.json'

export function randomName(): string {
  const first = randomArrayItem(firstName)
  const last = randomArrayItem(lastName.prefixes) + randomArrayItem(lastName.suffixes)

  if (Math.random() <= 0.10) {
    return `${first} ${randomArrayItem(middleName)} ${last}`
  }

  return `${first} ${last}`
}

export function randomRoll(max: number): number {
  return Math.floor(Math.random() * max) + 1
}

export function rollDice(dice: number, amount = 1): number[] {
  const rolls: number[] = []

  for (let i = 0; i < amount; i++) {
    rolls.push(randomRoll(+dice))
  }

  return rolls
}

const VALID_DICE_SIDES: DiceSide[] = [4, 6, 8, 10, 12, 20, 100]
const MAX_DICE_COUNT = 100

export function validateDiceExpression(diceExpr: string): boolean {
  const dicePattern = /^(\d+)d(\d+)$/
  const match = diceExpr.match(dicePattern)

  if (!match) return false

  const diceCount = parseInt(match[1]!)
  const diceSides = parseInt(match[2]!)

  if (diceCount <= 0 || diceCount > MAX_DICE_COUNT) return false
  if (!VALID_DICE_SIDES.includes(diceSides as DiceSide)) return false

  return true
}

export function parseDamageDice(damageDice?: string): { count: number, sides: number }[] {
  if (!damageDice) return []

  return damageDice
    .split(/[+\s]+/)
    .map(dice => dice.trim())
    .filter(dice => dice.length > 0)
    .filter(validateDiceExpression)
    .map((dice) => {
      const [count, sides] = dice.split('d')
      return { count: parseInt(count!), sides: parseInt(sides!) }
    })
}

type ModifierType = 'health' | 'ac'

function heal(type: ModifierType, row: InitiativeSheetRow, amount: number): void {
  const max = type === 'health' ? 'maxHealth' : 'maxAc'

  if (typeof row[type] === 'number' && typeof row[max] === 'number') {
    if (row[type] + amount > row[max]) row[type] = row[max]
    else row[type] = row[type] + amount
  }
}

function damage(type: ModifierType, row: InitiativeSheetRow, amount: number): void {
  const temp = type === 'health' ? 'tempHealth' : 'tempAc'

  if (typeof row[temp] === 'number') {
    if (row[temp] >= amount) row[temp] = row[temp] - amount
    else if (row[type]) {
      row[type] = row[type] - (amount - row[temp])
      row[temp] = 0
    }
  }
  else if (row[type]) row[type] = row[type] - amount
}

function temp(type: ModifierType, row: InitiativeSheetRow, amount: number): void {
  const temp = type === 'health' ? 'tempHealth' : 'tempAc'

  if (row[type] && row[type] > 0) row[temp] = amount
}

function override(type: ModifierType, row: InitiativeSheetRow, amount: number): void {
  const max = type === 'health' ? 'maxHealth' : 'maxAc'
  const old = type === 'health' ? 'maxHealthOld' : 'maxAcOld'

  if (row[type] && row[max]) {
    row[type] = amount < row[max]
      ? amount
      : amount - (row[max] - row[type])
  }
  else row[type] = amount

  row[old] = row[max] || amount
  row[max] = amount
}

function overrideReset(type: ModifierType, row: InitiativeSheetRow, amount: number): void {
  const max = type === 'health' ? 'maxHealth' : 'maxAc'
  const old = type === 'health' ? 'maxHealthOld' : 'maxAcOld'

  if (row[type] && row[max] && row[old]) {
    row[type] = row[old] < row[max]
      ? row[old]
      : row[old] - (row[max] - row[type])
  }

  row[old] = undefined
  row[max] = amount
}

export const hpFunctions = {
  heal: (row: InitiativeSheetRow, amount: number) => heal('health', row, amount),
  damage: (row: InitiativeSheetRow, amount: number) => damage('health', row, amount),
  temp: (row: InitiativeSheetRow, amount: number) => temp('health', row, amount),
  override: (row: InitiativeSheetRow, amount: number) => override('health', row, amount),
  overrideReset: (row: InitiativeSheetRow, amount: number) => overrideReset('health', row, amount),
}

export const acFunctions = {
  add: (row: InitiativeSheetRow, amount: number) => heal('ac', row, amount),
  remove: (row: InitiativeSheetRow, amount: number) => damage('ac', row, amount),
  temp: (row: InitiativeSheetRow, amount: number) => temp('ac', row, amount),
  override: (row: InitiativeSheetRow, amount: number) => override('ac', row, amount),
  overrideReset: (row: InitiativeSheetRow, amount: number) => overrideReset('ac', row, amount),
}

export function handleHpChanges(
  amount: number,
  type: HealthType,
  item: InitiativeSheetRow,
  allowNegative: boolean,
): { row: InitiativeSheetRow, toasts: ToastItem[] } {
  const toasts: ToastItem[] = []
  const row = { ...item }
  const noHp = typeof row.health === 'number' && row.health <= 0

  if (type === 'heal') {
    if (hasDeathSaves(row.type) && noHp) row.deathSaves = resetDeathSaves()

    heal('health', row, amount)
  }
  else if (type === 'temp') temp('health', row, amount)
  else if (type === 'override') override('health', row, amount)
  else if (type === 'override-reset') overrideReset('health', row, amount)
  else if (type === 'damage') {
    damage('health', row, amount)

    const downed = typeof row.health === 'number' && row.health <= 0

    // When the creature has 0hp and get damage, add 2 death save failures
    if (hasDeathSaves(row.type) && noHp) {
      if (!row.deathSaves) {
        row.deathSaves = {
          fail: [false, false, false],
          save: [false, false, false],
        }
      }

      row.deathSaves = addDeathSave(row.deathSaves, 'fail', 2)
    }

    if (row.concentration && !downed) {
      toasts.push({
        title: ['components.initiativeTable.concentration.title'],
        description: ['components.initiativeTable.concentration.text', { name: row.name }],
        variant: 'info',
      })
    }

    if (downed && (row.concentration || row.conditions.length)) {
      row.concentration = false
      row.conditions = []

      toasts.push({
        title: ['components.initiativeTable.downed.title', { name: row.name }],
        description: ['components.initiativeTable.downed.text', { name: row.name }],
        variant: 'info',
      })
    }
  }

  // when user is dies because of going to much in the negative hp
  const dead = (row.health && row.maxHealth && row.health < 0 && Math.abs(row.health) >= row.maxHealth)
  const { failed, saved } = row.deathSaves ? checkDeathSaves(row.deathSaves) : { failed: false, saved: false }

  if (dead || (failed && !saved)) {
    toasts.push({
      title: ['components.initiativeTable.died.title', { name: row.name }],
      description: ['components.initiativeTable.died.textMinHP', { name: row.name }],
      variant: 'info',
    })
  }

  if (!failed && saved) {
    toasts.push({
      title: ['components.initiativeTable.stable.title', { name: row.name }],
      description: ['components.initiativeTable.stable.textDeathSaves', { name: row.name }],
      variant: 'info',
    })
  }

  // when health is an negative number change it to 0
  if (!allowNegative && row.health && row.health < 0) row.health = 0

  return { row, toasts }
}

function hasDeathSaves(type: HomebrewType): boolean {
  return !['summon', 'lair'].includes(type)
}

function checkDeathSaves(deathSaves: DeathSaves): { failed: boolean, saved: boolean } {
  return {
    failed: deathSaves.fail.every(v => v === true) || false,
    saved: deathSaves.save.every(v => v === true) || false,
  }
}

function addDeathSave(deathSaves: DeathSaves, type: 'fail' | 'save' = 'fail', count: number = 1): DeathSaves {
  const updatedSaves = { ...deathSaves }
  const updated = [...updatedSaves[type]]
  let addedCount = 0

  for (let i = 0; i < updated.length && addedCount < count; i++) {
    if (!updated[i]) {
      updated[i] = true
      addedCount++
    }
  }

  updatedSaves[type] = updated as [boolean, boolean, boolean]
  return updatedSaves
}

function resetDeathSaves(): DeathSaves {
  return {
    fail: [false, false, false],
    save: [false, false, false],
  }
}

export const deathSavesFunctions = {
  hasDeathSaves,
  addDeathSave,
  checkDeathSaves,
  resetDeathSaves,
}

export function indexCorrect(rows: InitiativeSheetRow[]): InitiativeSheetRow[] {
  const sortedObjects = [...rows] // Make shallow copy

  // Sort the array based on the initiative value in descending order
  sortedObjects.sort((a, b) => {
    if (!a.initiative && !b.initiative) return 0
    if (!a.initiative) return 1
    if (!b.initiative) return -1
    return b.initiative - a.initiative
  })

  // Add the index to each object based on their sorted position
  return sortedObjects.map((obj, index) => ({ ...obj, index }))
}

export function getCurrentRowIndex(sheet: InitiativeSheet, id: string): number {
  return sheet.rows.findIndex(row => row.id === id)
}

function castToNumber(value: string | number): number | undefined {
  return isNaN(+value) ? undefined : +value
}

export function getHP(item: Open5eItem | Partial<InitiativeSheetRow> & { name: string }): number | undefined {
  if ('health' in item) {
    return isDefined(item.health) ? castToNumber(item.health) : undefined
  }

  if ('hit_points' in item) {
    return isDefined(item.hit_points) ? castToNumber(item.hit_points) : undefined
  }

  return undefined
}

export function getAC(item: Open5eItem | Partial<InitiativeSheetRow> & { name: string }): number | undefined {
  if ('ac' in item) {
    return isDefined(item.ac) ? castToNumber(item.ac) : undefined
  }

  if ('armor_class' in item) {
    return isDefined(item.armor_class) ? castToNumber(item.armor_class) : undefined
  }

  return undefined
}

export const createInitiativeRow = (
  formData: Open5eItem | Partial<InitiativeSheetRow> & { name: string },
  type: HomebrewType,
  encounterRows: number,
): InitiativeSheetRow => {
  const initiative = 'initiative' in formData ? Number(formData.initiative ?? -1) : -1
  const initiative_modifier = 'initiative_modifier' in formData ? Number(formData.initiative_modifier) : undefined
  const health = getHP(formData)
  const ac = getAC(formData)
  const baseArray: [boolean, boolean, boolean] = [false, false, false]

  const baseRow: Record<string, unknown> = {
    ...formData,
    id: crypto.randomUUID(),
    index: encounterRows + 1,
    conditions: [],
    type,
    initiative,
    initiative_modifier,
    note: '',
    health,
    ac,
    maxHealth: health,
    maxAc: ac,
    tempHealth: 0,
    tempAc: 0,
    concentration: false,
    deathSaves: { save: baseArray, fail: baseArray },
  }

  const allowedKeys = new Set(initiativeKeys)

  return Object.fromEntries(
    Object.entries(baseRow).filter(([key]) => allowedKeys.has(key)),
  ) as unknown as InitiativeSheetRow
}

export function hasMaxCharacters(sheet?: InitiativeSheet): boolean {
  return sheet ? sheet.rows.length >= 50 : false
}
