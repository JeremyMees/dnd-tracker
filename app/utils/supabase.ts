export async function sbQuery<T>(options: SbFetchOptions): Promise<SbQuery<T>> {
  const supabase = useSupabaseClient<DB>()

  const { table, select, page, perPage, filters, fuzzy, fields } = options
  const { eq, search, sortBy, sortDesc } = filters || {}

  let query = supabase
    .from(table)
    .select(select || '*', { count: 'estimated' })

  if (typeof page === 'number' && typeof perPage === 'number') {
    const { from, to } = sbRange(page, perPage)

    query = query.range(from, to)
  }

  if (sortBy) {
    query = query.order(sortBy, {
      ascending: sortDesc !== undefined ? !sortDesc : true,
      nullsFirst: false,
    })
  }

  if (eq) {
    query = query.eq(eq.field, eq.value)
  }

  if (search && fuzzy) {
    query = query.or(sbOrQuery(fields || ['title'], search))
  }

  const { data, error, count } = await query

  if (error) createError(error)

  return {
    data: data as T[],
    totalPages: perPage ? sbPages(count, perPage) : 1,
    count: count || 0,
  }
}

function sbOrQuery(keys: string[], search: string): string {
  let queryString = ''

  keys.forEach((key: string, i: number) => {
    const escapedSearch = search.replace(/([%_])/g, '\\$1')
    queryString += `${i ? ',' : ''}${key}.ilike.*${escapedSearch}*`
  })

  return queryString
}

export function sbRange(page: number, perPage: number): SbRange {
  const from = page ? page * perPage : 0
  const to = page ? from + perPage - 1 : perPage - 1

  return { from, to }
}

export function sbPages(count: number | null, perPage: number): number {
  return Math.ceil((count || 1) / perPage)
}

export function sbCount(key: string, obj: Record<string, any>): number {
  const value = obj[key]

  if (value && Array.isArray(value) && value.length) return value[0].count
  else return 0
}
