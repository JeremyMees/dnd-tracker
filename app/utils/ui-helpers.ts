import sanitizeHtml from 'sanitize-html'

export function sanitizeForm<T>(data: Record<string, any>, blacklist?: string[]): T {
  const garbage: string[] = blacklist || ['__init', 'isTrusted', '_vts']
  const sanitized: Record<string, any> = {}

  for (const key in data) {
    if (garbage.includes(key)) continue

    sanitized[key] = typeof data[key] === 'string'
      ? data[key].trim()
      : data[key]
  }

  return sanitized as T
}

export function scrollToId(id: string): void {
  const el = document.getElementById(id)

  if (el) {
    el.scrollIntoView({ behavior: 'smooth', block: 'end' })
  }
}

export function randomString(): string {
  return (Math.random() + 1).toString(36).substring(7)
}

export function randomColor(): string {
  return Math.floor(Math.random() * 16777215).toString(16)
}

export function togglePasswordInput(node: FormNode): void {
  node.props.suffixIcon = node.props.suffixIcon === 'tabler:eye' ? 'tabler:eye-closed' : 'tabler:eye'
  node.props.type = node.props.type === 'password' ? 'text' : 'password'
}

export function sortByNumber(a: number | any[], b: number | any[], acs: boolean): number {
  a = Array.isArray(a) ? a.length : a
  b = Array.isArray(b) ? b.length : b

  if (a == null) {
    return b == null ? 0 : 1
  }
  else if (b == null) {
    return -1
  }

  return acs ? a - b : b - a
}

export function sortByString(a: string, b: string, acs: boolean): number {
  a = a ?? ''
  b = b ?? ''

  return acs ? a.localeCompare(b) : b.localeCompare(a)
}

export function sortCreatedAt<T extends { created_at: string }>(arr: T[]): T[] {
  return arr.sort((a, b) => {
    return new Date(b.created_at).valueOf() - new Date(a.created_at).valueOf()
  })
}

export function focusInput({ node }: any): void {
  const id = node?.context?.id

  if (id) document.getElementById(id)?.focus()
}

export function homebrewIcon(type: HomebrewType): string {
  switch (type) {
    case 'summon': return 'tabler:wand'
    case 'npc': return 'tabler:user'
    case 'monster': return 'tabler:bat'
    case 'lair': return 'tabler:building-castle'
    default: return 'tabler:sword'
  }
}

export function homebrewBgColor(type: HomebrewType): string {
  switch (type) {
    case 'summon': return 'bg-tertiary'
    case 'npc': return 'bg-success'
    case 'monster': return 'bg-destructive'
    case 'lair': return 'bg-warning'
    default: return 'bg-primary'
  }
}

export function homebrewColor(type: HomebrewType): string {
  switch (type) {
    case 'summon': return 'text-tertiary'
    case 'npc': return 'text-success'
    case 'monster': return 'text-destructive'
    case 'lair': return 'text-warning'
    default: return 'text-primary'
  }
}

export function sanitizeHTML(dirty: string): string {
  return sanitizeHtml(dirty, {
    allowedTags: ['h1', 'h2', 'h3', 'p', 'a', 'ol', 'ul', 'li', 'blockquote', 'hr', 'mark', 'strong', 'em', 's'],
    allowedAttributes: {
      a: ['href', 'name', 'target', 'rel'],
    },
  }).replaceAll('<hr />', '<hr>')
}

export function formatDate(date: string | Date): string {
  const { locale } = useI18n()

  const formatter = new Intl.DateTimeFormat(
    locale.value,
    {
      year: '2-digit',
      month: '2-digit',
      day: '2-digit',
    },
  )

  return formatter.format(new Date(date))
}

export function isDefined<T>(value: T | null | undefined): value is T {
  return value !== null && value !== undefined
}

export function validateParamId(id: string | string[] | undefined): number {
  if (
    !id
    || typeof id !== 'string'
    || isNaN(+id)
  ) throw createError({ statusCode: 404, statusMessage: 'Id is not valid' })

  return +id
}
