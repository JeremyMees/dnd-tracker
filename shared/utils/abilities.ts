type Campaign = CampaignItem | CampaignFull | CampaignMinimal
type Encounter = EncounterItem

export const isOwner = (item: Campaign | Encounter, id: string) => item.created_by.id === id
export const isAdmin = (item: TeamMember[], id: string) => !!item.find(u => u.user.id === id && u.role === 'Admin')
export const isMember = (item: TeamMember[], id: string) => !!item.find(u => u.user.id === id)

export const isNotCampaignOwner = defineAbility((user: AuthUser, campaign: Campaign) => {
  return !isOwner(campaign, user.id)
})

export const isCampaignOwner = defineAbility((user: AuthUser, campaign: Campaign) => {
  return isOwner(campaign, user.id)
})

export const isCampaignAdmin = defineAbility((user: AuthUser, campaign: Campaign, strict = false) => {
  const owner = isOwner(campaign, user.id)
  const admin = isAdmin(campaign.team, user.id)

  return strict ? admin && !owner : admin || owner
})

export const isCampaignMember = defineAbility((user: AuthUser, campaign: Campaign, strict = false) => {
  const owner = isOwner(campaign, user.id)
  const admin = isAdmin(campaign.team, user.id)
  const member = isMember(campaign.team, user.id)

  return strict ? member && !owner && !admin : member || admin || owner
})

export const isEncounterOwner = defineAbility((user: AuthUser, encounter: Encounter) => {
  return isOwner(encounter, user.id)
})

export const isEncounterAdmin = defineAbility((user: AuthUser, encounter: Encounter, strict = false) => {
  const owner = isOwner(encounter, user.id)
  const admin = isAdmin(encounter.campaign?.team || [], user.id)

  return strict ? admin && !owner : admin || owner
})

export const isEncounterMember = defineAbility((user: AuthUser, encounter: Encounter, strict = false) => {
  const owner = isOwner(encounter, user.id)
  const admin = isAdmin(encounter.campaign?.team || [], user.id)
  const member = isMember(encounter.campaign?.team || [], user.id)

  return strict ? member && !owner && !admin : member || admin || owner
})

export const canUpdateEncounter = defineAbility((user: AuthUser, encounter: Encounter) => {
  const owner = isOwner(encounter, user.id)
  const admin = isAdmin(encounter.campaign?.team || [], user.id)
  const member = isMember(encounter.campaign?.team || [], user.id)

  if (admin) return true
  else if (owner && !member) return true
  else return false
})
