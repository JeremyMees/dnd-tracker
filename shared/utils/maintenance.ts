function urlMatches(url: string, patterns: string[]): boolean {
  for (const pattern of patterns) {
    const regex = new RegExp(`^${pattern.replace(/\*/g, '.*')}$`)
    if (regex.test(url)) return true
  }
  return false
}

export function isMaintenanceEnabled(): boolean {
  const { maintenanceMode } = useRuntimeConfig().public
  const type = typeof maintenanceMode

  const strValue = type === 'string'
    ? maintenanceMode.replace(/^"|"$/g, '')
    : maintenanceMode

  if (
    (type === 'string' && (strValue === 'true' || strValue === '1'))
    || (type === 'boolean' && maintenanceMode)
    || (type === 'number' && Number(maintenanceMode) === 1)
  ) return true
  else return false
}

export function isUnderMaintenance(path: string, exclude: string[] | null): boolean {
  if (exclude && urlMatches(path, exclude)) return false
  return true
}

export function throwMaintenanceError(): void {
  throw createError({
    statusCode: 503,
    statusMessage: 'Site is under maintenance',
  })
}
