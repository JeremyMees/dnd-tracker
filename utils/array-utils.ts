export function randomArrayItem<T>(arr: T[]): T {
  return arr[Math.floor(Math.random() * arr.length)]
}

export function randomArrayItems<T>(arr: T[], number = 1): T[] {
  const result: T[] = []
  const copy = [...arr]

  if (number >= arr.length) {
    return shuffleArray(arr)
  }

  for (let i = 0; i < number; i++) {
    const index = Math.floor(Math.random() * copy.length)
    result.push(copy.splice(index, 1)[0])
  }

  return result
}

export function searchArray<T>(arr: T[], key: keyof T, search: string): T[] {
  return arr
    .filter(item => (item[key] as string)
      .toLowerCase()
      .includes(search.toLowerCase().trim() || ''),
    )
}

export function toggleSelection<T extends { id: number }>(item: T, selected: T[]): void {
  const index: number = selected.findIndex(s => s.id === item.id)

  if (index === -1) selected.push(item)
  else selected.splice(index, 1)
}

export function shuffleArray<T>(arr: T[]): T[] {
  for (let i = 0; i < arr.length; i++) {
    const j = Math.floor(Math.random() * (i + 1))
    const temp = arr[i]
    arr[i] = arr[j]
    arr[j] = temp
  }

  return arr
}

export function splitArray<T>(arr: T[], size: number): T[][] {
  const result: T[][] = []

  for (let i = 0; i < size; i++) {
    result.push(arr.filter((_, index) => index % size === i))
  }

  return result
}
