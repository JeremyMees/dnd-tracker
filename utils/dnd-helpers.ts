import { firstName, lastName, middleName } from '~/constants/names.json'

export function randomName(): string {
  const first = randomArrayItem(firstName)
  const last = randomArrayItem(lastName.prefixes) + randomArrayItem(lastName.suffixes)

  if (Math.random() <= 0.10) {
    return `${first} ${randomArrayItem(middleName)} ${last}`
  }

  return `${first} ${last}`
}

export function randomRoll(max: number): number {
  return Math.floor(Math.random() * max) + 1
}

export function rollDice(dice: number, amount = 1): number[] {
  const rolls: number[] = []

  for (let i = 0; i < amount; i++) {
    rolls.push(randomRoll(+dice))
  }

  return rolls
}

type ModifierType = 'health' | 'ac'

function heal(type: ModifierType, row: InitiativeSheetRow, amount: number): void {
  const max = type === 'health' ? 'maxHealth' : 'maxAc'

  if (typeof row[type] === 'number' && typeof row[max] === 'number') {
    if (row[type] + amount > row[max]) row[type] = row[max]
    else row[type] = row[type] + amount
  }
}

function damage(type: ModifierType, row: InitiativeSheetRow, amount: number): void {
  const temp = type === 'health' ? 'tempHealth' : 'tempAc'

  if (typeof row[temp] === 'number') {
    if (row[temp] >= amount) row[temp] = row[temp] - amount
    else if (row[type]) {
      row[type] = row[type] - (amount - row[temp])
      row[temp] = 0
    }
  }
  else if (row[type]) row[type] = row[type] - amount
}

function temp(type: ModifierType, row: InitiativeSheetRow, amount: number): void {
  const temp = type === 'health' ? 'tempHealth' : 'tempAc'

  if (row[type] && row[type] > 0) row[temp] = amount
}

function override(type: ModifierType, row: InitiativeSheetRow, amount: number): void {
  const max = type === 'health' ? 'maxHealth' : 'maxAc'
  const old = type === 'health' ? 'maxHealthOld' : 'maxAcOld'

  if (row[type] && row[max]) {
    row[type] = amount < row[max]
      ? amount
      : amount - (row[max] - row[type])
  }
  else row[type] = amount

  row[old] = row[max] || amount
  row[max] = amount
}

function overrideReset(type: ModifierType, row: InitiativeSheetRow, amount: number): void {
  const max = type === 'health' ? 'maxHealth' : 'maxAc'
  const old = type === 'health' ? 'maxHealthOld' : 'maxAcOld'

  if (row[type] && row[max] && row[old]) {
    row[type] = row[old] < row[max]
      ? row[old]
      : row[old] - (row[max] - row[type])
  }

  row[old] = undefined
  row[max] = amount
}

export const hpFunctions = {
  heal: (row: InitiativeSheetRow, amount: number) => heal('health', row, amount),
  damage: (row: InitiativeSheetRow, amount: number) => damage('health', row, amount),
  temp: (row: InitiativeSheetRow, amount: number) => temp('health', row, amount),
  override: (row: InitiativeSheetRow, amount: number) => override('health', row, amount),
  overrideReset: (row: InitiativeSheetRow, amount: number) => overrideReset('health', row, amount),
}

export const acFunctions = {
  add: (row: InitiativeSheetRow, amount: number) => heal('ac', row, amount),
  remove: (row: InitiativeSheetRow, amount: number) => damage('ac', row, amount),
  temp: (row: InitiativeSheetRow, amount: number) => temp('ac', row, amount),
  override: (row: InitiativeSheetRow, amount: number) => override('ac', row, amount),
  overrideReset: (row: InitiativeSheetRow, amount: number) => overrideReset('ac', row, amount),
}

function hasDeathSaves(type: HomebrewType): boolean {
  return !['summon', 'lair'].includes(type)
}

function checkDeathSaves(deathSaves: DeathSaves): { failed: boolean, saved: boolean } {
  return {
    failed: deathSaves.fail.every(v => v === true) || false,
    saved: deathSaves.save.every(v => v === true) || false,
  }
}

function addDeathSave(deathSaves: DeathSaves, type: 'fail' | 'save' = 'fail', count: number = 1): DeathSaves {
  const updatedSaves = { ...deathSaves }
  const updated = [...updatedSaves[type]]
  let addedCount = 0

  for (let i = 0; i < updated.length && addedCount < count; i++) {
    if (!updated[i]) {
      updated[i] = true
      addedCount++
    }
  }

  updatedSaves[type] = updated as [boolean, boolean, boolean]
  return updatedSaves
}

function resetDeathSaves(): DeathSaves {
  return {
    fail: [false, false, false],
    save: [false, false, false],
  }
}

export const deathSavesFunctions = {
  hasDeathSaves,
  addDeathSave,
  checkDeathSaves,
  resetDeathSaves,
}
