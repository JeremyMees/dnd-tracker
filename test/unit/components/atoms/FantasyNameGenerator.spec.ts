import { mockNuxtImport, mountSuspended } from '@nuxt/test-utils/runtime'
import { describe, expect, it, vi } from 'vitest'
import * as dndHelpers from '~/utils/dnd-helpers'
import FantasyNameGenerator from '~/components/atoms/FantasyNameGenerator.vue'

interface Props {
  amount: number
}

const props: Props = {
  amount: 10,
}

const mockClipboard = vi.fn()

mockNuxtImport('useClipboard', () => () => ({
  copy: mockClipboard,
}))

describe('FantasyNameGenerator', async () => {
  it('Should match snapshot', async () => {
    vi.spyOn(dndHelpers, 'randomName').mockReturnValue('Test Name')

    const component = await mountSuspended(FantasyNameGenerator, { props })
    expect(component.html()).toMatchSnapshot()

    vi.restoreAllMocks()
  })

  it('Should render actions correctly', async () => {
    const component = await mountSuspended(FantasyNameGenerator, { props })
    const names = component.findAll('li')

    expect(names.length).toBe(props.amount)
  })

  it('Should generate new names when button is clicked', async () => {
    const component = await mountSuspended(FantasyNameGenerator, { props })

    const initialName = component.find('li').text()

    const generateButton = component.find('[data-test-generate]')
    await generateButton.trigger('click')
    await nextTick()

    const newName = component.find('li').text()

    expect(newName).not.toBe(initialName)
  })

  it('Should copy name when clicked', async () => {
    const component = await mountSuspended(FantasyNameGenerator, { props })

    const name = component.find('li')

    await name.trigger('click')
    await nextTick()

    expect(mockClipboard).toHaveBeenCalled()
  })
})
