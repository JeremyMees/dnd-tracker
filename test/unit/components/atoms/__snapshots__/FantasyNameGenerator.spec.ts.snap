// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FantasyNameGenerator > Should match snapshot 1`] = `
"<ol class="grid list-disc list-inside gap-x-6" style="grid-template-columns: repeat(2, minmax(0, 1fr));">
  <div class="flex flex-col gap-1">
    <li class="cursor-copy">Test Name</li>
    <li class="cursor-copy">Test Name</li>
    <li class="cursor-copy">Test Name</li>
    <li class="cursor-copy">Test Name</li>
    <li class="cursor-copy">Test Name</li>
  </div>
  <div class="flex flex-col gap-1">
    <li class="cursor-copy">Test Name</li>
    <li class="cursor-copy">Test Name</li>
    <li class="cursor-copy">Test Name</li>
    <li class="cursor-copy">Test Name</li>
    <li class="cursor-copy">Test Name</li>
  </div>
</ol>
<div class="flex w-full justify-end gap-2 items-center mt-4">
  <p class="text-muted-foreground text-2xs">pages.fantasyNameGenerator.tip</p><button data-test-generate="" class="btn-primary">actions.generate</button>
</div>"
`;
